examples/drivencavity.py

```diff
@@ -1,60 +1,141 @@
-    ns.add_field(('u', 'v'), domain.basis('std', degree=degree), shape=(domain.ndims,))
-    ns.add_field(('p', 'q'), domain.basis('std', degree=degree-1))
+    if not compatible:
+        ns.add_field(('u', 'v'), domain.basis('std', degree=degree), shape=(domain.ndims,))
+        ns.add_field(('p', 'q'), domain.basis('std', degree=degree-1)[1:])
+        ns.add_field('ψ', domain.basis('std', degree=2)[1:])
+    else:
+        ns.add_field(('u', 'v'), function.vectorize([
+            domain.basis('spline', degree=(degree, degree-1)),
+            domain.basis('spline', degree=(degree-1, degree))]))
+        ns.add_field(('p', 'q'), domain.basis('spline', degree=degree-1)[1:])
+        ns.add_field('ψ', domain.basis('spline', degree=degree)[1:])
     ns.σ_ij = '(∇_j(u_i) + ∇_i(u_j)) / Re - p δ_ij'
 
-    sqr = domain.boundary.integral('u_k u_k dS' @ ns, degree=degree*2)
-    sqr += domain.locate(ns.x, [(1,1)], weights=numpy.array([1.]), tol=1e-10).integral(ns.p**2)
-    cons = solver.optimize('u,p', sqr, droptol=1e-15)
+    res = domain.integral('∇_j(v_i) σ_ij dV' @ ns, degree=degree*2)
+    res += domain.integral('q ∇_k(u_k) dV' @ ns, degree=2*degree)
 
-    # project lid velocity separately to avoid Gibbs wiggles around boundary incompatibilities
-    sqr = domain.boundary['top'].integral('(u_0 - 1)^2 dS' @ ns, degree=degree*2)
-    lidcons = solver.optimize('u,', sqr, droptol=1e-15)
-    cons['u'] = numpy.choose(numpy.isnan(lidcons['u']), [lidcons['u'], cons['u']])
+    # strong enforcement of non-penetrating boundary conditions
+    sqr = domain.boundary.integral('(u_k n_k)^2 dS' @ ns, degree=degree*2)
+    cons = solver.optimize('u,', sqr, droptol=1e-15)
 
-    res = domain.integral('(∇_j(v_i) σ_ij + q ∇_k(u_k)) dV' @ ns, degree=degree*2)
-    with treelog.context('stokes'):
+    if strongbc:
+        # strong enforcement of tangential boundary conditions
+        sqr = domain.boundary['top,bottom'].integral('(u_0 - uwall_0)^2 dS' @ ns, degree=degree*2)
+        sqr += domain.boundary['right,left'].integral('(u_1 - uwall_1)^2 dS' @ ns, degree=degree*2)
+        tcons = solver.optimize('u,', sqr, droptol=1e-15)
+        cons['u'] = numpy.choose(numpy.isnan(cons['u']), [cons['u'], tcons['u']])
+    else:
+        # weak enforcement of tangential boundary conditions via Nitsche's method
+        ns.N = 5 * degree * nelems # Nitsche constant based on element size = 1/nelems
+        ns.nitsche_i = '(N v_i - (∇_j(v_i) + ∇_i(v_j)) n_j) / Re'
+        res += domain.boundary.integral('(nitsche_i (u_i - uwall_i) - v_i σ_ij n_j) dS' @ ns, degree=2*degree)
+
+    with log.context('stokes'):
         args0 = solver.solve_linear('u:v,p:q', res, constrain=cons)
         postprocess(domain, ns, **args0)
 
-    res += domain.integral('.5 (v_i ∇_j(u_i) - u_i ∇_j(v_i)) u_j dV' @ ns, degree=degree*3)
-    with treelog.context('navierstokes'):
+    res += domain.integral('v_i ∇_j(u_i) u_j dV' @ ns, degree=3*degree)
```

Comment:
The degree of this integral is higher than the other integrals in `res`. Does the performance improve if the other integrals are aligned?

Reply:
fixed